packages/trie/test/trie/prune.spec.ts

```diff
@@ -0,0 +1,59 @@
+import * as tape from 'tape'
+
+import { Trie /*decodeNode*/ } from '../../src'
+
+/*async function dumpTrieNodes(trie: any) {
+  const trieDB = trie.db._database
+  for (const key of trieDB.keys()) {
+    console.log(key)
+    console.log(decodeNode(trieDB.get(key)))
+  }
+}*/
+
+tape('Pruned trie tests', function (tester) {
+  const it = tester.test
+
+  it('should prune simple trie', async function (st) {
+    const trie = new Trie({ pruneTrie: true })
+    const key = Buffer.from('test')
+    await trie.put(key, Buffer.from('1'))
+    await trie.put(key, Buffer.from('2'))
+    await trie.put(key, Buffer.from('3'))
+    await trie.put(key, Buffer.from('4'))
+    await trie.put(key, Buffer.from('5'))
+    await trie.put(key, Buffer.from('6'))
+
+    st.equals((<any>trie.db)._database.size, 1, 'DB size correct')
+  })
+
+  it('should prune simple trie', async function (st) {
+    const trie = new Trie({ pruneTrie: true })
+    const key = Buffer.from('test')
+    await trie.put(key, Buffer.from('1'))
+    st.equals((<any>trie.db)._database.size, 1, 'DB size correct')
+
+    await trie.del(key)
+    st.equals((<any>trie.db)._database.size, 0, 'DB size correct')
+
+    await trie.put(key, Buffer.from('1'))
+    st.equals((<any>trie.db)._database.size, 1, 'DB size correct')
+  })
+
+  it('should prune trie with depth = 2', async function (st) {
+    const trie = new Trie({ pruneTrie: true })
+    // Create a Trie with
+    const keys = ['01', '02', '0103', '0104', '0105']
+    const values = ['00', '02', '03', '04', '05']
+
+    for (let i = 0; i < keys.length; i++) {
+      await trie.put(Buffer.from(keys[i], 'hex'), Buffer.from(values[i], 'hex'))
+    }
+
+    for (let i = 0; i < keys.length; i++) {
+      //console.log(await trie.get(Buffer.from(keys[i], 'hex')))
+    }
+
+    // TODO
+    st.end()
```

Comment:
In the end we should add a test on a random trie with 100 keys and 1000 random values which are all set to random keys (so some keys will be updated) and then finally also dropping 20 keys (deleting them). What we do here is: check if all the DB keys in the DB are reached (or the key is the root). If it cannot be reached (using `walkTrie` search using DFS/BFS) then the test has failed as we now have unreachable keys in the DB (which should be pruned).

Reply:
Wrote a test here that I think does what you suggest.

I definitely see that pruning has an effect when keys are updated.  If I run the test with `pruneTee: true`, the size of the DB doesn't grow when keys are updated, but it does if set to `pruneTree: false`.

Also when deleting.  with `pruneTree: true`,  the DB size shrinks when keys are deleted.  With `pruneTree: false`, it does not.

From your comment, though, I expected `walkTrie` to fail if `pruneTree: false` was set, but `walkTrie` runs either way.

Reply:
Yes, the `walkTrie` does not fail in this case, because it starts at the root (if we wold not be able to get the root from the db then the trie is definitely corrupted since we then cannot get any key/value pair). It also only checks the root (walkTrie needs you to reschedule tasks, see the `verifyPrunedTrie` test in newer commits)